src/app/clusters/service-area-server/service-area-server.h

```diff
@@ -0,0 +1,751 @@
+
+
+    ~Instance() override;
+
+    /**
+     * @brief Initialise the Service Area server instance.
+     * This function must be called after defining an Instance class object.
+     * @return an error if the given endpoint and cluster Id have not been enabled in zap or if the
+     * CommandHandler or AttributeHandler registration fails, else CHIP_NO_ERROR.
+     * This method also checks if the feature setting is valid, if invalid return value will be CHIP_ERROR_INVALID_ARGUMENT.
+     */
+    CHIP_ERROR Init();
+
+
+private:
+    Delegate * mDelegate;
+    EndpointId mEndpointId;
+    ClusterId mClusterId;
+
+    // Attribute Data Store
+    std::map<uint32_t, LocationStructureWrapper>       mSupportedLocations;  // class includes text storage for name field & utility functions
+    std::map<uint8_t, MapStructureWrapper>             mSupportedMaps;       // class includes text storage for name field & utility functions
+    std::list<uint32_t>                                mSelectedLocations;   // If this attribute lists more than one location, and the ListOrder feature is set to 1, 
+                                                                             // the device SHALL attempt to operate at the selected locations in the order they're listed in.
+    DataModel::Nullable<uint32_t>                      mCurrentLocation;     // may be volatile - callback to device to refresh value is provided
+    DataModel::Nullable<uint32_t>                      mEstimatedEndTime;    // may be volatile - callback to device to refresh value is provided
+    std::map<uint32_t, Structs::ProgressStruct::Type>  mProgressList;        // may be volatile - callback to device to refresh value is provided
+    BitMask<ServiceArea::Feature>                      mFeature;
+
+
+    // aliases for container types
+    using SupportedLocationPairType = decltype(mSupportedLocations)::value_type;
+    using SupportedLocationType     = decltype(mSupportedLocations)::mapped_type;
+    using SupportedMapPairType      = decltype(mSupportedMaps)::value_type; 
+    using SupportedMapType          = decltype(mSupportedMaps)::mapped_type;
+    using ProgressPairType          = decltype(mProgressList)::value_type; 
+    using ProgressType              = decltype(mProgressList)::mapped_type;
+
+
+    //*************************************************************************
+    // core functions
+
+    /**
+     * @brief Inherited from AttributeAccessInterface
+     * @return appropriately mapped CHIP_ERROR if applicable
+     */
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+
+    /**
+     * @brief Inherited from CommandHandlerInterface
+     */
+    void InvokeCommand(HandlerContext & ctx) override;
+
+
+    //*************************************************************************
+    // attribute readers
```

Comment:
For read, converting from the c++ containers to the json encoded output needs it's own particular handling, also one of the cases where the code walks through each list, rather than accessing a particular element with an Id value.